libfreerdp/core/info.c

```diff
@@ -43,6 +43,68 @@ static const char* const INFO_TYPE_LOGON_STRINGS[4] =
 	"Logon Extended Info"
 };
 
+/* This define limits the length of the strings in the label field. */
+#define MAX_LABEL_LENGTH 40
+static struct
+{
+	UINT32      flag;
+	const char* label;
+}  const info_flags[] =
+{
+	{INFO_MOUSE,                  "INFO_MOUSE"},
+	{INFO_DISABLECTRLALTDEL,      "INFO_DISABLECTRLALTDEL"},
+	{INFO_AUTOLOGON,              "INFO_AUTOLOGON"},
+	{INFO_UNICODE,                "INFO_UNICODE"},
+	{INFO_MAXIMIZESHELL,          "INFO_MAXIMIZESHELL"},
+	{INFO_LOGONNOTIFY,            "INFO_LOGONNOTIFY"},
+	{INFO_COMPRESSION,            "INFO_COMPRESSION"},
+	{INFO_ENABLEWINDOWSKEY,       "INFO_ENABLEWINDOWSKEY"},
+	{INFO_REMOTECONSOLEAUDIO,     "INFO_REMOTECONSOLEAUDIO"},
+	{INFO_FORCE_ENCRYPTED_CS_PDU, "INFO_FORCE_ENCRYPTED_CS_PDU"},
+	{INFO_RAIL,                   "INFO_RAIL"},
+	{INFO_LOGONERRORS,            "INFO_LOGONERRORS"},
+	{INFO_MOUSE_HAS_WHEEL,        "INFO_MOUSE_HAS_WHEEL"},
+	{INFO_PASSWORD_IS_SC_PIN,     "INFO_PASSWORD_IS_SC_PIN"},
+	{INFO_NOAUDIOPLAYBACK,        "INFO_NOAUDIOPLAYBACK"},
+	{INFO_USING_SAVED_CREDS,      "INFO_USING_SAVED_CREDS"},
+	{INFO_AUDIOCAPTURE,           "INFO_AUDIOCAPTURE"},
+	{INFO_VIDEO_DISABLE,          "INFO_VIDEO_DISABLE"},
+	{INFO_HIDEF_RAIL_SUPPORTED,   "INFO_HIDEF_RAIL_SUPPORTED"},
+};
+
+FREERDP_LOCAL char* rdp_info_package_flags_description(UINT32 flags)
+{
+	char* result;
+	size_t maximum_size = 0;
+	size_t i;
+	size_t size;
+
+	for (i = 0; i < ARRAYSIZE(info_flags); i ++)
+		maximum_size += strnlen(info_flags[i].label, MAX_LABEL_LENGTH) + 1;
+
+	result = calloc(maximum_size, sizeof(char));
+
+	if (!result)
+		return 0;
+
+	for (i = 0; i < ARRAYSIZE(info_flags); i ++)
+	{
+		if (info_flags[i].flag & flags)
+		{
+			strcat(result, info_flags[i].label);
+			strcat(result, "|");
```

Comment:
Potential OOB write if all flags are set (strcat adds a terminating \0 to the end of the string).